File: 207-course-schedule/course-schedule.js
var canFinish = function(numCourses, prerequisites) {
    
    const map = new Map(), visited = new Set(), visiting = new Set();
    for (let n = 0; n < numCourses; n++) {
        if (!map.has(n)) map.set(n, new Set());
    }
    for (let [course, preCourse] of prerequisites) {
        map.get(course).add(preCourse);
    }

    const hasCycle = (nde) => {
        if (visited.has(nde)) return false;
        if (visiting.has(nde)) return true;

        visiting.add(nde);
        for (let n of Array.from(map.get(nde))) {
            if (hasCycle(n)) return true;
        }
        visiting.delete(nde);
        visited.add(nde);
    }
    for (let n = 0; n < numCourses; n++) {
        if (hasCycle(n)) return false;
    }
    return true;
};

File: 207-course-schedule/course-schedule.swift
class Solution {
    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        var visited: Set<Int> = Set<Int>(), visiting: Set<Int> = Set<Int>()
        var adjList: [Int:Set<Int>] = [:]
        for i: Int in 0..<numCourses {
            if adjList[i] == nil { adjList[i] = Set<Int>()}
        }
        for pre: [Int] in prerequisites {
            adjList[pre[0]]!.insert(pre[1])
        }

        func hasCycle(_ nde: Int) -> Bool {
            if visited.contains(nde) { return false }
            if visiting.contains(nde) { return true }
            visiting.insert(nde)
            for child: Int in adjList[nde]! {
                if hasCycle(child) { return true }
            }
            visiting.remove(nde)
            visited.insert(nde)
            return false
        }
        for i: Int in 0..<numCourses {
            if hasCycle(i) { return false }
        }
        return true
    }
}